concurrency-primer.tex

```diff
@@ -443,52 +464,54 @@ \subsection{Compare and swap}
 It allows us to conditionally exchange a value \emph{if} its previous value matches some expected one.
 In \clang{} and \cplusplus{}, \textsc{CAS} resembles the following,
 if it were executed atomically:
-\begin{cppcode}
-template <typename T>
-bool atomic<T>::compare_exchange_strong(
-    T& expected, T desired)
+\begin{ccode}
+/* A is an atomic type. C is the non-atomic type corresponding to A */
+bool atomic_compare_exchange_strong(A* obj, C* expected, C desired)
 {
-    if (*this == expected) {
-        *this = desired;
+    if (memcmp(obj, expected, sizeof(*object)) == 0) {
+        memcpy(obj, &desired, sizeof(*object));
         return true;
+    } else {
+        memcpy(expected, obj, sizeof(*object));
+        return false;
     }
-    expected = *this;
-    return false;
 }
-\end{cppcode}
+\end{ccode}
 
 \begin{samepage}
 \noindent The \cpp|_strong| suffix may leave you wondering if there is a corresponding ``weak'' \textsc{CAS}.
 Indeed, there is. However, we will delve into that topic later in \secref{spurious-llsc-failures}.
 \end{samepage}
 
-Let's say we have some long-running task that we might want to cancel.
-We'll give it three states: \textit{idle}, \textit{running},
-and \textit{cancelled}, and write a loop that exits when it is cancelled.
-\begin{cppcode}
-enum class TaskState : int8_t {
-    Idle, Running, Cancelled
-};
+Once threads are created in the thread pool as workers, they will continuously search for jobs to do.
+Jobs are taken from the tail of job queue.
+To claim a job without it being taken by another worker halfway through, we need to atomically change the pointer to the last job. Otherwise the last job is under races.
```

Comment:
Should provide the example result of races with `-fsanitize=thread -g` enabled.

Reply:
Thread sanitizer currently does not support C11 thread `[1]` `[2]` `[3]`*. Using functions in `threads.h` with `-fsanitize=thread` results in segmentation fault in thread sanitizer runtime. I will provide a minimum modification in git diff format in this subsection to run the example with Tsan.

`[1]`: https://clang.llvm.org/docs/ThreadSanitizer.html
`[2]`: https://github.com/google/sanitizers/wiki/ThreadSanitizerCppManual#faq
`[3]`: https://github.com/intel/libipt/commit/f24d71340f70dcdc2f738333bc6c09cec6509a81
*: Both documents [1] and [2] don't clearly specify that Tsan doesn't support C11 thread yet. By most-recent commit confirming this [3] and C++11 thread still requiring rebuilding libc++ with Tsan, I believe C11 is not supported in Tsan yet.